### PR TITLE
[Core] Install memory monitor callback to kill worker when memory usage is above threshold

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

Why are these changes needed?

The Ray-level OOM killer preemptively kills a worker process when the node is under memory pressure. This PR leverages the memory monitor from https://github.com/ray-project/ray/pull/27017 and supersedes https://github.com/ray-project/ray/pull/26962 to kill worker processes when the system is running low on memory. The node manager implements the callback in the memory monitor and kills the worker process with the newest task. It evicts only one worker at a time and enforces that by tracking the last evicted worker. If the eviction is still in progress it will not evict another worker even if the memory usage is above the threshold.

This PR is a no-op since the monitor is disabled by default.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Ran the tests in linux container / Ray CI that is using cgroup v1

RFC:
https://docs.google.com/document/d/1AG1Nx2znKZLsld92V1hvGIcsY-iSIzeuzhbs74w-70g/edit#heading=h.17dss3b9evbj

### CODE REVIEW

**Comment by rkooo567** [comment]
Can you also link the design doc for this feature btw?

**Comment by clarng** [comment]
> Can you also link the design doc for this feature btw (I remember there was one before)?

added to pr

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -2860,6 +2871,50 @@ void NodeManager::PublishInfeasibleTaskError(const RayTask &task) const {
   }
 }
 
+MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
+  return [this](bool is_usage_above_threshold) {
```
Comment (scv119) [inline src/ray/raylet/node_manager.cc:2875]: we need some inline comments on the policy we pick the worker to kill. also should we unit test this?

Comment (scv119) [reply src/ray/raylet/node_manager.cc:2875]: (can be done later): we can have a separate component to pick a worker to kill.

**Comment by scv119** [comment]
also, we do have a flag to turn this off right?

**Review by scv119 — COMMENTED** [review]
mostly nits and questions.

**Comment by clarng** [comment]
> also, we do have a flag to turn this off right?

It is off right now, controlled by the periodic runner period which is defaulted to zero

**Review by scv119 — APPROVED** [review]
might need merge master again.

**Pull request merged**